File: CCC/ccc20s3/src/ccc20s3.java
import java.util.*;
import java.io.*;

public class ccc20s3 {
    /* 
     * Copy-pasting code is NOT cool! Please do not copy and paste my code as a submission to DMOJ.
     * github.com/jdabtieu/competitive-programming
     */
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer in;

    static final long MMOD = 100000000183L;
    static final int PRIME = 101;
    static int ll;
    static long hash = 0;

    static int[] cnt = new int[26];
    static int[] cnt2 = new int[26];
    
    static String small, big;

    public static void main(String[] args) throws IOException {
        small = readLine();
        big = readLine();
        ll = small.length();
        if (ll > big.length()) {
            System.out.println(0);
            return;
        }
        long power = 1;
        for (int i = 0; i < ll; i++) power = power * PRIME % MMOD;
        Set<Long> found = new HashSet<>();
        for (int i = 0; i < ll; i++) {
            cnt[small.charAt(i) - 'a']++;
            cnt2[big.charAt(i) - 'a']++;
            hash = (hash * PRIME + big.charAt(i) - 'a') % MMOD;
        }
        if (check()) {
            found.add(hash);
        }
        for (int i = ll; i < big.length(); i++) {
            cnt2[big.charAt(i - ll) - 'a']--;
            cnt2[big.charAt(i) - 'a']++;
            hash = (hash * PRIME + big.charAt(i) - 'a') % MMOD;
            hash -= (big.charAt(i - ll) - 'a') * power % MMOD;
            if (hash < 0) hash += MMOD;
            if (check()) {
                found.add(hash);
            }
        }
        System.out.println(found.size());
    }
    
    static boolean check() {
        for (int i = 0; i < 26; i++) if (cnt[i] != cnt2[i]) return false;
        return true;
    }

    static String next() throws IOException {
        while (in == null || !in.hasMoreTokens())
            in = new StringTokenizer(br.readLine());
        return in.nextToken();
    }

    static long readLong() throws IOException {
        return Long.parseLong(next());
    }

    static int readInt() throws IOException {
        return Integer.parseInt(next());
    }

    static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }

    static char readChar() throws IOException {
        return next().charAt(0);
    }

    static String readLine() throws IOException {
        return br.readLine();
    }
}

File: CCC/ccc20s3/src/ccc20s3.cpp
#include <bits/stdc++.h>
#include <unistd.h>
#include <sys/syscall.h>
#pragma GCC optimize ("Ofast")

#define BUF_SIZE 65536

char ibuf[BUF_SIZE];
char obuf[BUF_SIZE];
int iPtr = 0;
int maxPtr = 0;
int oPtr = 0;
void Read() {maxPtr = syscall(SYS_read, 0, &ibuf, BUF_SIZE);iPtr = 0;}
void Write() {syscall(SYS_write, 1, &obuf, oPtr);oPtr = 0;}
void pc(char c) {if (oPtr >= BUF_SIZE) Write();obuf[oPtr++] = c;}
void _pi(int n) {if (n == 0) return;_pi(n / 10);pc((char) (n % 10 + '0'));}
void pi(int num) {if (num < 0) {pc('-');_pi(-num);} else if (num == 0) {pc('0');} else {_pi(num);}}
char sc() {if (iPtr >= maxPtr) Read();return ibuf[iPtr++];}
int si() {int x = 0;char c;while ((c = sc()) <= ' ');int neg = 0;if (c == '-') {neg = 1;c = sc();}do {x = x * 10 + (c - '0');} while ((c = sc()) > ' ');return neg ? -x : x;}
unsigned int su() {int x = 0;char c;while ((c = sc()) <= ' ');do {x = x * 10 + (c - '0');} while ((c = sc()) > ' ');return x;}

#define pb push_back
#define pii pair<int, int>
typedef unsigned long long llu;
typedef long long ll;
typedef unsigned int uint;
typedef __uint128_t lllu;
typedef __int128_t lll;
/*
 * Copy-pasting code is NOT cool! Please do not copy and paste my code as a submission to DMOJ.
 * github.com/jdabtieu/competitive-programming
 */
using namespace std;

const ll MMOD = 1125899839733759LL;
const int PRIME = 7507;

int l;
ll has = 0;

int cnt[26], cnt2[26];

string small, big;

bool check() {
    for (int i = 0; i < 26; i++) if (cnt[i] != cnt2[i]) return false;
    return true;
}

int main() {
    cin.sync_with_stdio(0);
    cin.tie(0);
    cin >> small >> big;
    l = small.length();
    if (l > big.length()) {
        printf("0\n");
        return 0;
    }
    ll power = 1;
    for (int i = 0; i < l; i++) power = power * PRIME % MMOD;
    unordered_set<ll> found;
    for (int i = 0; i < l; i++) {
        cnt[small[i] - 'a']++;
        cnt2[big[i] - 'a']++;
        has = (has * PRIME + big[i] - 'a') % MMOD;
    }
    
    if (check()) {
        found.insert(has);
    }
    
    for (int i = l; i < big.length(); i++) {
        cnt2[big[i - l] - 'a']--;
        cnt2[big[i] - 'a']++;
        has = (has * PRIME + big[i] - 'a') % MMOD;
        has -= (big[i - l] - 'a') * power % MMOD;
        if (has < 0) has += MMOD;
        if (check()) {
            found.insert(has);
        }
    }
    printf("%d\n", found.size());
}
